### PR TITLE
Multi level

### PR DESCRIPTION
**Pull request by rohit-mp**
Added sample code structure. Suggest changes if any.

### CODE REVIEW

**Comment by shanthanu9** [comment]
The following could be a better format for config file

```
<number of levels>
<L1 cache policy> <cs> <bs> <sa>
<L2 cache policy> <cs> <bs> <sa>
...
```

**Comment by shanthanu9** [comment]
Introducing two modes could be good: debug and non-debug.

**Comment by rohit-mp** [comment]
Runs into a supposedly infinite loop. Needs debugging.

File: cache.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    protected:
+        Cache(long long cacheSize, long long blockSize, long long setASsociativity);
+        long long cacheSize;
+        long long blockSize;
+        long long setAssociativity;
+        long long numberOfSets;
+        int offsetSize;
+        int indexSize;
+        long long* cacheBlocks;
+        virtual ~Cache() = 0;
```
Comment (shanthanu9) [inline cache.h]: ~Cache() should deallocate cacheBlocks

File: cache.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    protected:
+        Cache(long long cacheSize, long long blockSize, long long setASsociativity);
+        long long cacheSize;
+        long long blockSize;
+        long long setAssociativity;
+        long long numberOfSets;
+        int offsetSize;
+        int indexSize;
+        long long* cacheBlocks;
```
Comment (shanthanu9) [inline cache.h]: Also sure you want policy writer to be given direct access to cacheBlocks? Seems unnecesary

File: plru.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                cacheBlocks[setNumber*(setAssociativity-1) + node/2] = false;
+            }
+            else{
+                cacheBlocks[setNumber*(setAssociativity-1) + node/2] = true;
+            }
+            node /= 2;
+        }
+    }
+
+    PLRU::~PLRU(){}
```
Comment (shanthanu9) [inline plru.cpp]: Should dealloc tree.

File: cache.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+        long long getHits();
+        long long getMisses();
+        float getHitRate();
+
+        virtual long long getBlockToReplace(long long address) = 0;
+        virtual void update(long long blockToReplace, int status) = 0;
+
+    protected:
+        Cache(long long cacheSize, long long blockSize, long long setASsociativity);
```
Comment (shanthanu9) [inline cache.h]: Typo in third parameter

File: main.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+using namespace std;
+using namespace std::chrono;
+
+#define ll long long
+
+Cache* createCacheInstance(string& policy, ll cs, ll bs, ll sa){
+    
+    // check validity here and exit if invalid
+    if(!strcmp(policy.c_str(), "plru")){
```
Comment (shanthanu9) [inline main.cpp]:
```suggestion
    if(policy == "plru"){
```
This should be fine. (Btw, checkout : [https://www.rttr.org/](https://www.rttr.org/))

Comment (shanthanu9) [inline main.cpp]:
```suggestion
    if(policy == "plru"){
```
This should be fine. (Btw, checkout : [https://www.rttr.org/](https://www.rttr.org/))

File: cache.h (hunk 2 above)
Comment (rohit-mp) [reply cache.h]: He atleast needs to have read access and anyways we'll be dealing with pointer to the cacheBlocks by using getters, all to the same effect.

**Pull request merged**